### PR TITLE
Dependencies update for C# driver.

### PR DESCRIPTION
**Pull request by farost**
## What is the goal of this PR?

We add all the needed dependencies for building the new C# driver for TypeDB.

## What are the changes implemented in this PR?

This PR contains bazel's csharp rules' pulling commands and additional bazel rules for building the native C# library used in the driver and its automatic tests. 

This PR **will be updated** after we finish the deployment preparations work on later stages of the development, so there are still some TODOs left. The current state of this PR is just needed for referencing from the main [C# driver's PR](https://github.com/vaticle/typedb-driver/pull/609). 

I've put a **status: blocked** label for explicitness, but I'm not sure what's the correct process, so feel free to correct me.

### CODE REVIEW

**Comment by farost** [comment]
Self review incoming in several hours.

File: builder/swig/csharp.bzl [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+# TODO: Generalise and move the same functions (used by java, python, csharp)
```
Comment (farost) [inline builder/swig/csharp.bzl:18]: Not sure if it's still relevant. There are some similarities, but such refactoring could make the readability worse. Will think on the deployment stage.

File: builder/swig/csharp.bzl [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        quote_includes = lib_compilation_context.quote_includes,
+        system_includes = depset(
+            [],
+            transitive = [lib_compilation_context.system_includes],
+        )
+    )
+
+    return [
+        DefaultInfo(files = depset([wrap_cxx, wrap_csharp])),
+        OutputGroupInfo(
```
Comment (farost) [inline builder/swig/csharp.bzl:82]: This is a useful technique to save separate sets of files for future usage. I had troubles with separating C# and CXX files from each other to compile both source and test libraries, and it was a perfect solution!

File: builder/swig/csharp.bzl [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+)
+
+
+def _csharp_native_library(ctx):
+    return [
+        DefaultInfo(
+            files = depset([ctx.file.native_lib]),
+            runfiles = ctx.runfiles(files = [ctx.file.native_lib])
+        ),
+        DotnetAssemblyCompileInfo(
```
Comment (farost) [inline builder/swig/csharp.bzl:139]: We need to put DotnetAssembly*Info to build our test (specifically BDD test) targets, which require this information for dependencies.

File: builder/swig/csharp.bzl [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+                # TODO: move http certificate/encryption libraries into arguments
+                "@vaticle_bazel_distribution//platform:is_windows": ["ntdll.lib", "secur32.lib", "crypt32.lib", "ncrypt.lib"],
+                "//conditions:default": [],
+            }),
+        )
+
+    native_lib_name_root = "typedb_driver"
+    native_lib_name = "{}_native".format(native_lib_name_root)
+
+    # TODO: On Mac, it's enough to pass native_lib_name_root (lib and .dylib are added by the rule)
```
Comment (farost) [inline builder/swig/csharp.bzl:201]: Will make an additional PR for it.

File: builder/swig/csharp.bzl [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+    def swig_cc_binary(shared_lib_filename):
+        # name doesn't accept select() either
+        native.cc_binary(
+            name = shared_lib_filename,
+            deps = [lib, swig_wrapper_name],
+            srcs = [swig_wrapper_name],
+            linkshared = True,
+            linkopts = select({
+                # TODO: move http certificate/encryption libraries into arguments
```
Comment (farost) [inline builder/swig/csharp.bzl:192]: Will check it out in an additional PR for deployment.

File: builder/csharp/deps.bzl [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+def deps(use_patched_version=False):
+    http_archive(
+        name = "rules_dotnet",
+        sha256 = "09021aa1d8a63395cd072e384cd560612e713f35a448da43a7e89087787aadc0",
+        strip_prefix = "rules_dotnet-0.13.0",
+        url = "https://github.com/bazelbuild/rules_dotnet/releases/download/v0.13.0/rules_dotnet-v0.13.0.tar.gz",
+    )
```
Comment (dmitrii-ubskii) [inline builder/csharp/deps.bzl:26]: ⛔️

File: WORKSPACE [hunk 7]
```diff
@@ -49,6 +49,12 @@ kotlin_repositories()
 load("@io_bazel_rules_kotlin//kotlin:core.bzl", "kt_register_toolchains")
 kt_register_toolchains()
 
+# Load //builder/csharp
+load("//builder/csharp:deps.bzl", dotnet_deps = "deps")
+dotnet_deps()
+load("@rules_dotnet//dotnet:repositories.bzl", "rules_dotnet_dependencies")
+rules_dotnet_dependencies()
+
```
Comment (dmitrii-ubskii) [inline WORKSPACE:57]: Do we actually have any .NET targets in dependencies? This should only appear in the repos that do.

File: builder/swig/csharp.bzl [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+        ),
+        DotnetAssemblyRuntimeInfo(
+            name = "{}".format(ctx.file.native_lib.basename),
+            version = "",
+            deps = depset(),
+            nuget_info = [],
+            libs = [],
+            native = [ctx.file.native_lib],
+            direct_deps_depsjson_fragment = [],
+        )]
```
Comment (dmitrii-ubskii) [inline builder/swig/csharp.bzl:156]: * How many of these empty args are defaulted?
* `"{}".format(...)` does nothing, `File.basename` is already a `string`

**Review by dmitrii-ubskii — APPROVED** [review]
Generally approved provided the couple minor comments are addressed

**Comment by dmitrii-ubskii** [comment]
Looks like you have some unsigned commits. You'll want to amend them with a signature before you merge.

File: builder/csharp/deps.bzl (hunk 6 above)
Comment (farost) [reply builder/csharp/deps.bzl:26]: Fixed!

File: WORKSPACE (hunk 7 above)
Comment (farost) [reply WORKSPACE:57]: Looks like you're right, we don't need it here.

File: builder/swig/csharp.bzl (hunk 8 above)
Comment (farost) [reply builder/swig/csharp.bzl:156]: * Not a single one, I was inserting it while getting errors.
* I know, but it looked better for me to have all the rules names in a similar format with .format. I can remove it if it disturbs a reader too much.

Comment (farost) [reply builder/swig/csharp.bzl:156]: Approved irl

**Pull request merged**